lua-api-crates/table-funcs/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+// Note that we don't use a HashMap since we want to keep the order of the tables, which
+// can be useful in some cases
+fn deep_extend<'lua>(
+    lua: &'lua Lua,
+    (array_of_tables, behavior): (Vec<Table<'lua>>, Option<ConflictMode>),
+) -> mlua::Result<Table<'lua>> {
+    let mut tbl_vec: Vec<(LuaValue, LuaValue)> = vec![];
+    for table in array_of_tables {
+        for pair in table.pairs::<LuaValue, LuaValue>() {
+            let (key, value) = pair?;
+            tbl_vec.push((key, value));
+        }
+    }
+    let tbl_len = tbl_vec.len();
+    // note we might allocate a bit too much here, but in many use cases we will be correct
+    let tbl: Table<'lua> = lua.create_table_with_capacity(0, tbl_len)?;
+
+    match behavior {
+        Some(ConflictMode::Keep) => {
+            for (key, value) in tbl_vec {
+                if !tbl.contains_key(key.clone())? {
+                    tbl.set(key, value)?;
+                } else if let LuaValue::Table(t) = value {
+                    let inner_tbl = deep_extend(
+                        lua,
+                        (vec![tbl.get(key.clone())?, t], Some(ConflictMode::Keep)),
+                    )?;
+                    tbl.set(key, inner_tbl)?;
+                }
+            }
+        }
+        // default behavior is to keep last set value
+        Some(ConflictMode::Force) | None => {
+            for (key, value) in tbl_vec {
+                if !tbl.contains_key(key.clone())? {
+                    tbl.set(key, value)?;
+                } else if let LuaValue::Table(t) = value {
+                    let inner_tbl = deep_extend(
+                        lua,
+                        (vec![tbl.get(key.clone())?, t], Some(ConflictMode::Force)),
+                    )?;
+                    tbl.set(key, inner_tbl)?;
+                } else {
+                    tbl.set(key, value)?;
+                }
+            }
+        }
+        Some(ConflictMode::Error) => {
+            for (key, value) in tbl_vec {
+                if !tbl.contains_key(key.clone())? {
+                    tbl.set(key, value)?;
+                } else if let LuaValue::Table(t) = value {
+                    let inner_tbl = deep_extend(
+                        lua,
+                        (vec![tbl.get(key.clone())?, t], Some(ConflictMode::Keep)),
+                    )?;
+                    tbl.set(key, inner_tbl)?;
+                } else {
+                    return Err(mlua::Error::runtime(format!(
+                        "The key {} is in more than one of the tables.",
+                        key.to_string()?
+                    )));
+                }
+            }
+        }
+    };
+
+    Ok(tbl)
+}
+
+fn clone<'lua>(lua: &'lua Lua, table: Table<'lua>) -> mlua::Result<Table<'lua>> {
+    let table_len = table.clone().pairs::<LuaValue, LuaValue>().count();
+    let res: Table<'lua> = lua.create_table_with_capacity(0, table_len)?;
```

Comment:
The way that the full length is computed essentially clones the table and discards it, which feels more expensive than is perhaps necessary.  I would just create an empty one and fill it as we go; it feels like the lesser of two evils, and would be the same as using `{}` to create a new table in pure lua code anyway.


```suggestion
    let res: Table<'lua> = lua.create_table()?;
```